### PR TITLE
AA8K-10 - Rewrite code and  write test.

### PR DESCRIPTION
**Pull request by Bartek-Figat**
Rewrite User Schema to TS and rewrite test jest to TS. I faced an error when was trying run test.
ReferenceError: You are trying to `import` a file after the Jest environment has been torn down.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,14 +4,24 @@
...
   "main": "index.js",
   "scripts": {
-    "test": "echo \"WARNING: no test specified; Change 'exit 0' to 'exit 1' at the beggining of configuring this!\" && exit 0"
+    "test": "jest"
   },
   "keywords": [],
   "author": "",
   "license": "ISC",
   "dependencies": {
+    "@types/mongodb": "^3.5.27",
```
Comment (hayuna) [inline package.json]: Do we still use this types? If no - remove it, if yes, types should be stored in devDependencies

File: src/__tests__user__/insert.test.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import mongoose from "mongoose";
+import User, { IUser } from "../models/User";
+
+describe("User model", () => {
+  beforeAll(async () => {
+    await mongoose.connect("mongodb://localhost:27017", {
```
Comment (hayuna) [inline src/__tests__user__/insert.test.ts:6]: I think we could use `test.env` to provide these configuration

File: src/models/User.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import mongoose, { Schema, Document } from "mongoose";
+
+export interface IUser extends Document {
+  githubId: string;
```
Comment (hayuna) [inline src/models/User.ts:5]: What if we are working with non-developer which use google to authorize himself?

File: src/crud__operation.ts [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    return await collection.deleteOne(query);
+  } catch (error) {
+    console.log(error);
+  } finally {
+    await client.close();
+  }
+};
+
+//Delet many documents
+export const deleteManyDoc = async (query: object) => {
```
Comment (hayuna) [inline src/crud__operation.ts]: Better name will be `deleteManyDocuments`
Please replace this type `object` in something more related

File: src/crud__operation.ts [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+  const client = new MongoClient(lockalConnection, dbOptions);
```
Comment (hayuna) [inline src/crud__operation.ts]: We are using mongoose, is it still used?

File: src/crud__operation.ts [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+  }
+};
+
+//Replace one document
+export const replaceOneDoc = async (
+  query: object,
+  replacement: object,
+  options: object
+) => {
+  //=====Query
```
Comment (hayuna) [inline src/crud__operation.ts]: If this code is ready for review, please remove your comments. Comments should be used only if logic is super complex and we need to explain other developers behaviour

File: src/__tests__user__/insert.test.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  });
+
+  it("Should throw validation errors", () => {
+    const user = new User();
+
+    expect(user.validate).toThrow();
+  });
+
+  it("Should save a user", async () => {
+    expect.assertions(3);
```
Comment (hayuna) [inline src/__tests__user__/insert.test.ts]: I have never used expect.assertions, could you explain this in the comment?

File: src/models/User.ts (hunk 3 above)
Comment (Bartek-Figat) [reply src/models/User.ts:5]: As far as i know. For now we chose using only github auth.

**Comment by hayuna** [comment]
BTW

[Boo tests failed](https://github.com/Programmers-Only-Group/auth/pull/6/checks?check_run_id=1083337280#step:6:1)

File: package.json (hunk 1 above)
Comment (Bartek-Figat) [reply package.json]: Concept change after last call, i will delete this library.

File: src/crud__operation.ts (hunk 4 above)
Comment (Bartek-Figat) [reply src/crud__operation.ts]: This file will by deleted. We not longer using this functions.

File: src/crud__operation.ts (hunk 5 above)
Comment (Bartek-Figat) [reply src/crud__operation.ts]: Yes, mongoose will by main library for CRUD operations.

**Comment by hayuna** [comment]
> ReferenceError: You are trying to `import` a file after the Jest environment has been torn down.

Tell us about your investigation to resolve this :)

File: src/models/User.ts (hunk 3 above)
Comment (hayuna) [reply src/models/User.ts:5]: Not exactly, we are using github for developers and google for testers/UI/UX/ and rest of users who aren't dev

File: src/crud__operation.ts (hunk 6 above)
Comment (Bartek-Figat) [reply src/crud__operation.ts]: This file will by deleted.

File: src/__tests__user__/insert.test.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+      username: "Test username",
+      token: "Test token",
+    });
+    const spy = jest.spyOn(user, "save");
+
+    await user.save();
+
+    expect(spy).toHaveBeenCalled();
+
+    expect(user).toMatchObject({
```
Comment (Przemocny) [inline src/__tests__user__/insert.test.ts]: here should be perfect match, not expect.any(String) - wrong test logic

File: src/__tests__user__/insert.test.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+      githubId: expect.any(String),
+      username: expect.any(String),
+      token: expect.any(String),
+    });
+
+    expect(user.token).toBe("Test token");
```
Comment (Przemocny) [inline src/__tests__user__/insert.test.ts]: this should be in .toMatchObject

File: src/__tests__user__/insert.test.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+  afterAll(async () => {
+    await User.findOneAndDelete({ token: "Test token" });
+    await mongoose.connection.close();
+  });
+
+  it("Should throw validation errors", async () => {
+    const user = new User();
+
+    expect(user.validate).toThrow();
```
Comment (Przemocny) [inline src/__tests__user__/insert.test.ts]: .toThrow is 1/2

next half is what error message precisely do you expect

**Review by Przemocny — CHANGES_REQUESTED** [review]
little fixes

File: src/models/User.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+  token: string;
+}
+
+const UserSchema: Schema = new Schema({
+  githubId: { type: String, required: true, unique: true },
+  username: { type: String, required: true },
+  token: { type: String, required: true },
+});
+
+// Export the model and return your IUser interface
```
Comment (hayuna) [inline src/models/User.ts]: Is this comment helpful for anyone?

File: src/__tests__user__/insert.test.ts [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+  it("Should save a user", async () => {
+    const user: IUser = new User({
+      githubId: "Test github Id",
+      username: "Test username",
+      token: "Test token",
+    });
+    const spy = jest.spyOn(user, "save");
+    const isgithubId = user.githubId;
+    const isUsername = user.username;
+    const istoken = user.token;
```
Comment (hayuna) [inline src/__tests__user__/insert.test.ts:34]: What is the point of these variables?

Comment (Bartek-Figat) [reply src/__tests__user__/insert.test.ts:34]: I'll check if the value is what it expects. I used Object.is to check for strict equality.

**Pull request closed without merging**